src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java

```diff
@@ -0,0 +1,239 @@
+    }
+
+    @Override
+    public int getSlotLimit(int slot)
+    {
+        return 64;
+    }
+
+    @Override
+    public boolean isItemValid(int slot, @NotNull ItemStack stack)
+    {
+        return stack.getItem().canFitInsideContainerItems();
+    }
+
+    @Override
+    public void setStackInSlot(int slot, @NotNull ItemStack stack) {
+        validateSlotIndex(slot);
+        if (!isItemValid(slot, stack)) throw new RuntimeException("Invalid stack " + stack + " for slot " + slot + ")");
+        NonNullList<ItemStack> itemStacks = getItemList();
+        itemStacks.set(slot, stack);
+        setItemList(itemStacks);
+    }
+
+    protected NonNullList<ItemStack> getItemList()
+    {
+        CompoundTag rootTag = BlockItem.getBlockEntityData(this.stack);
+        if ((cachedTag != null && !cachedTag.equals(rootTag)) || rootTag == null)
+            itemStacksCache = refreshItemList(rootTag);
+        return itemStacksCache;
+    }
+
+    protected NonNullList<ItemStack> refreshItemList(CompoundTag rootTag)
+    {
+        NonNullList<ItemStack> itemStacks = NonNullList.withSize(getSlots(), ItemStack.EMPTY);
+        if (rootTag != null && rootTag.contains("Items", 9)) {
+            ContainerHelper.loadAllItems(rootTag, itemStacks);
+        }
+        cachedTag = rootTag;
+        return itemStacks;
+    }
+
+    protected void setItemList(NonNullList<ItemStack> itemStacks)
+    {
+        CompoundTag existing = BlockItem.getBlockEntityData(this.stack);
+        CompoundTag rootTag = ContainerHelper.saveAllItems(existing == null ? new CompoundTag() : existing, itemStacks);
+        BlockItem.setBlockEntityData(this.stack, BlockEntityType.SHULKER_BOX,
+                rootTag);
+        cachedTag = rootTag;
+    }
+
+    @Override
+    @NotNull
+    public <T> LazyOptional<T> getCapability(@NotNull Capability<T> cap, @Nullable Direction side)
+    {
+        return CapabilityItemHandler.ITEM_HANDLER_CAPABILITY.orEmpty(cap, this.holder);
+    }
+
+    @Mod.EventBusSubscriber(modid = "forge")
+    private static class AttachmentHandler
+    {
+        private static final Set<Item> SHULKER_ITEMS = Set.of(Items.SHULKER_BOX,
+                Items.BLACK_SHULKER_BOX,
+                Items.BLUE_SHULKER_BOX,
+                Items.BROWN_SHULKER_BOX,
+                Items.CYAN_SHULKER_BOX,
+                Items.GRAY_SHULKER_BOX,
+                Items.GREEN_SHULKER_BOX,
+                Items.LIGHT_BLUE_SHULKER_BOX,
+                Items.LIGHT_GRAY_SHULKER_BOX,
+                Items.LIME_SHULKER_BOX,
+                Items.MAGENTA_SHULKER_BOX,
+                Items.ORANGE_SHULKER_BOX,
+                Items.PINK_SHULKER_BOX,
+                Items.PURPLE_SHULKER_BOX,
+                Items.RED_SHULKER_BOX,
+                Items.WHITE_SHULKER_BOX,
+                Items.YELLOW_SHULKER_BOX);
+
+        @SubscribeEvent
+        public static void listenCapabilitiesAttachment(AttachCapabilitiesEvent<ItemStack> event) {
```

Comment:
This should use `initCapabilities` in the Item class, not the event.

Reply:
As discussed multiple times, shulker boxes use normal BlockItems.

Reply:
The only way to make this use initCapabilities would be to make an -unused ShulkerBoxBlockItem class and patch it in for shulkers, which seemed like overkill to me. If this is the preferred solution, please let me know

Reply:
No, it's not. You *could* change  `public static void listenCapabilitiesAttachment(AttachCapabilitiesEvent<ItemStack> event) `  to `@ApiInternal public static ICapabilityPrvider getVanillaShulkerInventory() ` and call it from `IForgeItem.initCapabilities`'s default implementation.

Also, don't use a static set. Or make the set based on the delegate.

Reply:
As discussed on discord already, we *don't* want to use delegates here, because if someone registry-replaces the shulker boxes, we don't want to attach out capability.

Reply:
And I disagree, because if you implement it in initCaps, then they can opt out via NOT calling  super.

Reply:
That doesn't solve the issue: imagine we attach this in `initCapabilities` checking equality with a delegate. Someone else replaces shulker boxes with their own normal `BlockItem`. We've now attached a capability to that BlockItem, even though they never asked us to.

Reply:
That is their problem. Opt out vs Opt In.

Reply:
Alright, I should rephrase: if we do this with delegates, it becomes a breaking change. If you're alright with that, I'll go ahead and switch it to delegates then.

Reply:
By that definition this is a breaking change by adding it at all.
So no, you're wrong. 
And i've already told you what to do. So do it. I've already told this to my triage team about 5 times.